packages/backend-output-storage/src/stack_metadata_output_storage_strategy.ts

```diff
@@ -61,32 +66,75 @@ export class StackMetadataBackendOutputStorageStrategy
           `Metadata entry for ${keyName} at version ${existingMetadataEntry.version} already exists. Cannot add another entry for the same key at version ${version}.`
         );
       }
-    } else {
-      this.stack.addMetadata(keyName, {
-        version,
-        stackOutputs: Lazy.list({
-          produce: () => Array.from(listsMap ? listsMap.keys() : []),
-        }),
-      });
     }
 
-    Object.entries(backendOutputEntry.payload).forEach(([listName, value]) => {
-      if (!listsMap) {
-        listsMap = new Map();
-        this.lazyListValueMap.set(keyName, listsMap);
-      }
-      let outputList = listsMap.get(listName);
+    this.addOrUpdateMetadata(
+      existingMetadataEntry,
+      keyName,
+      backendOutputEntry as BackendOutputEntry
+    );
 
-      if (outputList) {
-        outputList.push(value);
-      } else {
-        outputList = [value];
-        listsMap.set(listName, outputList);
+    Object.entries(backendOutputEntry.payload ?? []).forEach(
+      ([listName, value]) => {
+        if (!value) {
+          return;
+        }
+        if (!listsMap) {
+          listsMap = new Map();
+          this.lazyListValueMap.set(keyName, listsMap);
+        }
+        let outputList = listsMap.get(listName);
 
-        new CfnOutput(this.stack, listName, {
-          value: Lazy.string({ produce: () => JSON.stringify(outputList) }),
-        });
+        if (outputList) {
+          outputList.push(value);
+        } else {
+          outputList = [value];
+          listsMap.set(listName, outputList);
+
+          new CfnOutput(this.stack, listName, {
+            value: Lazy.string({ produce: () => JSON.stringify(outputList) }),
+          });
+        }
       }
-    });
+    );
   };
+
+  /**
+   * Add or update metadata entry.
+   * @param existingMetadataEntry - The existing metadata entry.
+   * @param keyName - The key name.
+   * @param backendOutputEntry - The backend output entry.
+   */
+  private addOrUpdateMetadata(
+    existingMetadataEntry:
+      | { version: string; stackOutputs: string[] }
+      | undefined,
+    keyName: string,
+    backendOutputEntry: BackendOutputEntry
+  ) {
+    /*
+     * If there's only one bucket and not the bucket is not set as default, then it only has 'buckets' in the stackOutputs.
+     * So we need to manually add the bucketName and storageRegion to the stackOutputs.
+     * As long as there's a bucket, the bucketName, storageRegion, and buckets will always be present in the stackOutputs.
+     */
+    const stackOutputs = Object.keys(backendOutputEntry.payload).includes(
+      'buckets'
+    )
+      ? ['buckets', 'bucketName', 'storageRegion']
+      : Object.keys(backendOutputEntry.payload);
+
```

Comment:
I may need to re-write the comment. The idea is...

- not default bucket -> `appendToBackendOutputList` -> `metaData` has `[buckets]`
- default bucket -> `addBackendOutputEntry` and `appendToBackendOutputList` -> `metaData` has `['buckets', 'bucketName', 'storageRegion']`

There's no problem if there's a default bucket, but if there's no default bucket (We allow to have no default bucket if there's only one bucket), the `metaData` is lack of `[bucketName', 'storageRegion']`

Reply:
I think I understand the problem now. You don't know the best time to add the top level `bucketName` and `storageRegion` to the metadata as they exist today, since if a bucket is marked default, that should be added to the metadata, but if not (and it's the only bucket) you'd have to delay the addition until you are sure that there exists only one bucket.

I think a little less hacky solution for this would be to get hold of `outputStorageStrategy` from the `getInstanceProps` in the storage factory. You can then use the `outputStorageStrategy.addBackendOutputEntry()` in the aspect that you are creating at the storage bucket nested stack level, at which point you have the complete picture of all the buckets. This way you can also avoid [creating the CFNOutputs manually in the aspect ](https://github.com/aws-amplify/amplify-backend/pull/1742/files/db34c500c02e0994a7e62d8b043b44bce83affd5#diff-27a0769d9fccd4e1743973fa17e8ef386c1ac148f35c41fce90bbfe11d0837afR98-R103)

@sobolk thoughts?

Reply:
Refactored in 179a1bb9c8e48b89f1e449a16fba3c57a1fcca47, but I didn't figure out an elegant way to run Aspect only once. Any suggestions? @edwardfoyle @sobolk

Reply:
oh, I was trying to @Amplifiyer instead 😂